subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchableHierarchies.java

```diff
@@ -110,6 +118,7 @@ public SnapshotHierarchy removeUnwatchableContentOnBuildFinished(SnapshotHierarc
         if (!shouldWatchUnsupportedFileSystems(watchMode)) {
             newRoot = removeUnwatchableFileSystems(newRoot, invalidator);
         }
+        watchableHierarchiesSinceLastBuildFinish.clear();
```

Comment:
If we only clear it on build start, then the list would contain watchable hierarchies since the last build start. Since we only want to capture hierarchies since the last build finish, we need to clear it here or somehow make sure we don't capture anything from build start to build finish.

We could also clear it additionally on build start, though I couldn't come up with a good reason why we should.